In [36]:
import pandas as pd
import numpy as np
import mysql.connector

In [4]:
import pandas as pd

customers = pd.read_csv(r"D:\Personal\Data Science\Data Science\Revision\End to End Project\emsoft\customers.csv")
subscriptions = pd.read_csv(r"D:\Personal\Data Science\Data Science\Revision\End to End Project\emsoft\subscriptions.csv")
events = pd.read_csv(r"D:\Personal\Data Science\Data Science\Revision\End to End Project\emsoft\events.csv")

print(customers.shape)
print(subscriptions.shape)
print(events.shape)


(1000, 5)
(941, 6)
(2411, 5)


In [5]:
customers.head()

customer_id signup_date     segment country  is_enterprise
0       C0001  2023-01-04  Enterprise      IN           True
1       C0002  2023-01-14         SMB      CA          False
2       C0003  2023-01-04         SMB      IN          False
3       C0004  2023-03-19         SMB      CA          False
4       C0005  2023-03-25         NaN      IN          False

In [6]:
subscriptions.head()

subscription_id customer_id  start_date    end_date  monthly_price    status
0          S00001       C0001  2023-01-10         NaN            499    active
1          S00002       C0002  2023-01-12         NaN             49    active
2          S00003       C0003  2023-01-08  2023-02-14            149  canceled
3          S00004       C0004  2023-01-21         NaN            149    active
4          S00005       C0005  2023-04-03         NaN             99    active

In [7]:
events.head()

event_id customer_id   event_type  event_date    source
0  E000001       C0001       signup  2023-03-10   organic
1  E000002       C0001       signup  2023-03-10   organic
2  E000003       C0002       signup  2023-03-27       ads
3  E000004       C0002  trial_start  2023-03-28       ads
4  E000005       C0003       signup  2023-02-19  referral

STEP 2: Initial Data Exploration

In [38]:
## Missing values
print(customers.isnull().sum())
print("==========\n")
print(subscriptions.isnull().sum())
print("==========\n")
print(events.isnull().sum())

customer_id        0
signup_date       36
segment          243
country            0
is_enterprise      0
dtype: int64

subscription_id      0
customer_id          0
start_date           0
end_date           718
monthly_price        0
status               0
dtype: int64

event_id       0
customer_id    0
event_type     0
event_date     0
source         0
dtype: int64


In [43]:
# Data type inspection
print(customers.dtypes)
print("==========\n")
print(subscriptions.dtypes)
print("==========\n")
print(events.dtypes)

customer_id      object
signup_date      object
segment          object
country          object
is_enterprise      bool
dtype: object

subscription_id            object
customer_id                object
start_date         datetime64[ns]
end_date                   object
monthly_price               int64
status                     object
dtype: object

event_id       object
customer_id    object
event_type     object
event_date     object
source         object
dtype: object


In [44]:
# Unique values sanity

print(customers['segment'].value_counts(dropna=False))
print("==========\n")
print(subscriptions['status'].value_counts(dropna=False))
print("==========\n")
print(events['event_type'].value_counts(dropna=False))


segment
SMB           261
Enterprise    250
Mid-Market    246
NaN           243
Name: count, dtype: int64

status
active      718
canceled    223
Name: count, dtype: int64

event_type
signup         1094
trial_start     676
activated       404
churned         237
Name: count, dtype: int64


STEP 3: Validate SQL ingestion results

In [18]:
import mysql.connector

conn = mysql.connector.connect(
    host='localhost',
    user='root',
    password='sususuju3',
    database='emergence_saas'
)


In [25]:
from sqlalchemy import create_engine

engine = create_engine(
    "mysql+mysqlconnector://root:sususuju3@localhost/emergence_saas"
)



In [46]:
print(pd.read_sql("SELECT COUNT(*) AS cnt FROM customers_clean", engine))
print(pd.read_sql("SELECT COUNT(*) AS cnt FROM subscriptions_clean", engine))
print(pd.read_sql("SELECT COUNT(*) AS cnt FROM events_clean", engine))

    cnt
0  1000
   cnt
0  941
    cnt
0  2411


In [48]:
print(customers.shape[0])
print(subscriptions.shape[0])
print(events.shape[0])

1000
941
2411


In [27]:
# Row count validation


sql_counts = {
    "customers": pd.read_sql("SELECT COUNT(*) AS cnt FROM customers_clean", engine),
    "subscriptions": pd.read_sql("SELECT COUNT(*) AS cnt FROM subscriptions_clean", engine),
    "events": pd.read_sql("SELECT COUNT(*) AS cnt FROM events_clean", engine)
}

sql_counts





{'customers':     cnt
 0  1000,
 'subscriptions':    cnt
 0  941,
 'events':     cnt
 0  2411}

In [30]:
sql_counts = {
    name: pd.read_sql(query, engine).iloc[0, 0]
    for name, query in {
        "customers": "SELECT COUNT(*) FROM customers_clean",
        "subscriptions": "SELECT COUNT(*) FROM subscriptions_clean",
        "events": "SELECT COUNT(*) FROM events_clean"
    }.items()
}

sql_counts


{'customers': np.int64(1000),
 'subscriptions': np.int64(941),
 'events': np.int64(2411)}

In [49]:
# Churn definition validation (CRITICAL)

churn_event_customers = (
    events[events['event_type'] == 'churned']['customer_id']
    .nunique()
)

canceled_subscription_customers = (
    subscriptions[subscriptions['status'] == 'canceled']['customer_id']
    .nunique()
)

churn_event_customers, canceled_subscription_customers


(237, 223)

In [50]:
# SQL-side churn mismatch validation (CROSS-CHECK)

query = """
SELECT COUNT(DISTINCT e.customer_id) AS mismatch_customers
FROM events_clean e
LEFT JOIN subscriptions_clean s
  ON e.customer_id = s.customer_id
WHERE e.event_type = 'churned'
  AND (s.status IS NULL OR s.status != 'canceled');
"""

pd.read_sql(query, engine)

mismatch_customers
0                 184

In [51]:
# Numeric sanity check using NumPy

np.percentile(
    subscriptions['monthly_price'].dropna().astype(float),
    [0, 25, 50, 75, 100]
)


array([ 49.,  79., 149., 499., 699.])

STEP 4: Validate Key Metrics

In [32]:
# MRR sanity check

sql_mrr = pd.read_sql("""
SELECT
    DATE_FORMAT(start_date, '%Y-%m-01') AS month,
    SUM(monthly_price) AS mrr
FROM subscriptions_clean
GROUP BY month
ORDER BY month
""", engine)

sql_mrr.head()



month      mrr
0  2023-01-01  64742.0
1  2023-02-01  76084.0
2  2023-03-01  80691.0
3  2023-04-01  17832.0

In [33]:
# validating in pandas
subscriptions['start_date'] = pd.to_datetime(subscriptions['start_date'], errors='coerce')

pandas_mrr = (
    subscriptions
    .dropna(subset=['monthly_price'])
    .groupby(subscriptions['start_date'].dt.to_period('M'))
    ['monthly_price']
    .sum()
)

pandas_mrr.head()


start_date
2023-01    64742
2023-02    76084
2023-03    80691
2023-04    17832
Freq: M, Name: monthly_price, dtype: int64

STEP 5: Sanity Check

In [35]:
# Churn mismatch check

churn_events = events[events['event_type'] == 'churned']['customer_id'].nunique()
canceled_subs = subscriptions[subscriptions['status'] == 'canceled']['customer_id'].nunique()

print(churn_events, canceled_subs)


237 223


Multiple churn-related counts were observed due to differing data grains.
Event-based churn identified 237 unique customers, while subscription cancellations identified 223 customers.
A subset of 184 customers had churn events without corresponding canceled subscriptions, highlighting inconsistencies between behavioral and billing data.
For this reason, subscription status was used as the source of truth for churn metrics, and churn events were used only for funnel analysis.